datafusion/core/tests/parquet/schema_adapter.rs

```diff
@@ -0,0 +1,171 @@
+use datafusion::datasource::physical_plan::{
+    FileScanConfig, ParquetExec, SchemaAdapter, SchemaAdapterFactory, SchemaMapper,
+};
+use datafusion::physical_plan::{collect, Statistics};
+use datafusion::prelude::SessionContext;
+
+use datafusion::datasource::listing::PartitionedFile;
+use parquet::arrow::ArrowWriter;
+use tempfile::TempDir;
+
+#[tokio::test]
+async fn can_override_schema_adapter() {
+    // Create several parquet files in same directoty / table with
+    // same schema but different metadata
+    let tmp_dir = TempDir::new().unwrap();
+    let table_dir = tmp_dir.path().join("parquet_test");
+    fs::DirBuilder::new().create(table_dir.as_path()).unwrap();
+    let f1 = Field::new("id", DataType::Int32, true);
+
+    let file_schema = Arc::new(Schema::new(vec![f1.clone()]));
+    let filename = "part.parquet".to_string();
+    let path = table_dir.as_path().join(filename.clone());
+    let file = fs::File::create(path.clone()).unwrap();
+    let mut writer = ArrowWriter::try_new(file, file_schema.clone(), None).unwrap();
+
+    let ids = Arc::new(Int32Array::from(vec![1i32]));
+    let rec_batch = RecordBatch::try_new(file_schema.clone(), vec![ids]).unwrap();
+
+    writer.write(&rec_batch).unwrap();
+    writer.close().unwrap();
+
+    let location = Path::parse(path.to_str().unwrap()).unwrap();
+    let metadata = std::fs::metadata(path.as_path()).expect("Local file metadata");
+    let meta = ObjectMeta {
+        location,
+        last_modified: metadata.modified().map(chrono::DateTime::from).unwrap(),
+        size: metadata.len() as usize,
+        e_tag: None,
+        version: None,
+    };
+
+    let partitioned_file = PartitionedFile {
+        object_meta: meta,
+        partition_values: vec![],
+        range: None,
+        statistics: None,
+        extensions: None,
+    };
+
+    let f1 = Field::new("id", DataType::Int32, true);
+    let f2 = Field::new("extra_column", DataType::Utf8, true);
+
+    let schema = Arc::new(Schema::new(vec![f1.clone(), f2.clone()]));
+
+    // prepare the scan
+    let parquet_exec = ParquetExec::new(
+        FileScanConfig {
+            object_store_url: ObjectStoreUrl::local_filesystem(),
+            file_groups: vec![vec![partitioned_file]],
+            statistics: Statistics::new_unknown(&schema),
+            file_schema: schema,
+            projection: None,
+            limit: None,
+            table_partition_cols: vec![],
+            output_ordering: vec![],
+        },
+        None,
+        None,
+        Default::default(),
+    )
+    .with_schema_adapter_factory(Arc::new(TestSchemaAdapterFactory {}));
+
+    let session_ctx = SessionContext::new();
+    let task_ctx = session_ctx.task_ctx();
+    let read = collect(Arc::new(parquet_exec), task_ctx).await.unwrap();
+
+    let expected = [
+        "+----+--------------+",
+        "| id | extra_column |",
+        "+----+--------------+",
+        "| 1  | foo          |",
```

Comment:
It might help future readers if you left some comments explaining that the point of the test was to inject a column that doesn't appear in any of the files